docs/proposals/004-initialization.md

```diff
@@ -0,0 +1,317 @@
+
+### Single argument constructor call
+
+Call to a constructor with a single argument is always treated as a syntactic sugar of type cast:
+```csharp
+int x = int(1.0f); // is treated as (int) 1.0f;
+MyType y = MyType(arg); // is treated as (MyType)arg;
+MyType x = MyType(y); // equivalent to `x = y`.
+```
+
+The compiler will attempt to resolve all type casts using type coercion rules, if that failed, will fall back to resolve it as a constructor call.
+
+### Initialization List
+
+Slang allows initialization of a variable by assigning it with an initialization list. 
+Generally, Slang will always try to resolve initialization list coercion as if it is an explicit constructor invocation.
+For example, given:
+```csharp
+S obj = {1,2};
+```
+Slang will try to convert the code into:
+```csharp
+S obj = S(1,2);
+```
+
+Following the same logic, an empty initializer list will translate into a default-initialization:
+```csharp
+S obj = {};
+// equivalent to:
+S obj = S();
+```
+
+Note that initializer list of a single argument still translates directly into a constructor call and not a type cast. For example:
+```csharp
+void test()
+{
+  MyType t = {1};
+  // translates to:
+  // MyType t = MyType.__init(1);
+  // which is not
+  // MyType t = MyType(t)
+  // or
+  // MyType t = (MyType)t;
+}
+```
+
+If the above code passes type check, then it will be used as the way to initialize `obj`.
+
+If the above code does not pass type check, Slang continues to check if `S` meets the standard of a "legacy C-style struct` type.
+A type is a "legacy C-Style struct" if all of the following conditions are met:
+- It is a user-defined struct type or a basic scalar, vector or matrix type, e.g. `int`, `float4x4`.
+- It does not contain any explicit constructors defined by the user.
+- All its members have higher or equal visibility than the type.
+- All its members are legacy C-Style structs or arrays of legacy C-style structs.
+In such case, we perform a legacy "read data" style consumption of the initializer list, so that the following behavior is valid:
+
+```csharp
+struct Inner { int x; int y; };
+struct Outer { Inner i; Inner j; }
+
+Outer o = {1, 2, 3}; // Initializes `o` into `{ Inner{1,2}, Inner{3,0} }`.
+```
+
+If the type is not a legacy C-Style struct, Slang should produce an error.
+
+
+Examples
+-------------------
+```csharp
+
+// Assume everything below is public unless explicitly declared.
+
+struct Empty {}
+void test()
+{
+  Empty s0 = {}; // Works, `s` is considered initialized.
+  Empty s1; // `s1` is considered initialized bcause `Empty` is `IDefaultInitializable`.
+}
+
+struct CLike {int x; int y; }
+void test1()
+{
+  CLike c0; // `c0` is uninitialized.
+  CLike c1 = {}; // initialized with legacy initializaer list logic, `c1` is now `{0,0}`.
+  CLike c2 = {1}; // initialized with legacy initializaer list logic, `c1` is now `{1,0}`.
```

Comment:
Looking at this example, so we still zero-initialize members?

Reply:
And we still want to support legacy initializaer list?

Reply:
Yes, that is a still fallback if ctor match failed.

Reply:
I can verify that this is already supported by current slang.